src/generator/AutoRest.Ruby/Templates/ServiceClientTemplate.cshtml

```diff
@@ -79,6 +79,62 @@ module @Settings.Namespace
     end
 
     @EmptyLine
+    #
+    # Makes a request and returns the body of the response
+    # @@param method [Symbol] with any of the following values :get, :put, :post, :patch, :delete   
+    # @@param path [String] the path, relative to {api_endpoint}
+    # @@param options [Hash{String=>String}] specifying any request options like :body
+    # @@return [Hash{String=>String}] containing the body of the response
+    #
+    def make_request(method, path, options = {})
+      result = make_request_async(method, path, options).value!
+      result.response.body.to_s.empty? ? nil : JSON.load(result.response.body)
+    end
+    @EmptyLine
+
+    #
+    # Makes a request asynchronously
+    # @@param method [Symbol] with any of the following values :get, :put, :post, :patch, :delete
+    # @@param path [String] the path, relative to {api_endpoint}
+    # @@param options [Hash{String=>String}] specifying any request options like :body
+    # @@return [Concurrent::Promise] Promise object which holds the HTTP response.
+    #
+    def make_request_async(method, path, options = {})
+      fail ArgumentError, 'method is nil' if method.nil?
+      fail ArgumentError, 'path is nil' if path.nil?
+
+      @EmptyLine
+      request_headers= REQUEST_HEADERS
+      request_headers.merge!({'accept-language' => @@accept_language}) unless @@accept_language.nil?
+
+      @EmptyLine
+      request_url= options[:api_endpoint] || @@base_url
+      
+      @EmptyLine
+      options=options.merge({
```

Comment:
minor: I'd keep all the header related merging in subsequent lines: 

```
temp_headers = REQUEST_HEADERS.merge({'accept-language' => @accept_language}) unless @accept_language.nil?
options.merge!({headers: temp_headers.merge(options[:headers] || {})})
options.merge!({credentials: @@credentials}) unless @@credentials.nil?
```

I am not sure but seems like this should be equivalent to what you are trying or may be not. Just a thought as i did spend time in understanding it and i hope i did.